File: 프로그래머스/3/132266. 부대복귀/부대복귀.js
// source : 각 부대원이 위치한 지역
// roads : 길 정보
// 길이 없는 경우에는 -1 return 
function solution(n, roads, sources, destination) {
    var answer = [];
    var roadinfo = new Map();
    for(var i=1; i<=n; i++) roadinfo.set(i,[]);
    
    for(var [from,to] of roads){
        roadinfo.set(from,[...roadinfo.get(from),to]);
        roadinfo.set(to,[...roadinfo.get(to),from]);
    }
        var data = new Array(n).fill(Infinity);
        var queue = [];
        const visited = new Array(n+1).fill(false);
        var min = Infinity;
        queue.push([destination,0]);
        var queueindex=0;
        while(queue.length>queueindex){
            var [loc,count] = queue[queueindex];
            visited[loc]=true;
            queueindex++; 
            data[loc-1] = Math.min(count,data[loc-1]);
            for(var nextLoc of roadinfo.get(loc)){
                if(!visited[nextLoc]){
                    queue.push([nextLoc,count+1]);
                }
            }
        }
    for(var s of sources){
        answer.push(data[s-1]===Infinity?-1:data[s-1])
    }

    return answer;
}

File: 프로그래머스/3/132266. 부대복귀/부대복귀.java
import java.util.*;
class Solution {
    public int[] solution(int n, int[][] roads, int[] sources, int destination) {
        
        HashMap<Integer,List<Integer>> mapInfo = new HashMap<>();
        for(int[] road : roads){
            mapInfo
                .computeIfAbsent(road[0],k->new ArrayList<>())
                .add(road[1]);
            mapInfo
                .computeIfAbsent(road[1],k->new ArrayList<>())
                .add(road[0]);
        }
        Integer[] distance = new Integer[n+1];
        Arrays.fill(distance,Integer.MAX_VALUE);
        distance[destination]=0;
        Queue<Integer[]> queue = new LinkedList<>();
        queue.add(new Integer[]{destination,0});
        while(!queue.isEmpty()){
            Integer[] info = queue.poll();
            distance[info[0]]=Math.min(info[1],distance[info[0]]);
            for(Integer nextNode : mapInfo.get(info[0])){
                if(distance[nextNode]>info[1]+1)
                queue.add(new Integer[]{nextNode,info[1]+1});
            }
        }

        int[] answer = new int[sources.length];
        for(int i=0; i<sources.length; i++){
            answer[i] = distance[sources[i]]==Integer.MAX_VALUE?-1:distance[sources[i]];
        }
        return answer;
    }
}